modules/messages/src/lib.rs

```diff
@@ -499,13 +500,27 @@ pub mod pallet {
 				});
 
 				// if some new messages have been confirmed, reward relayers
-				T::DeliveryConfirmationPayments::pay_reward(
+				let actually_rewarded_relayers = T::DeliveryConfirmationPayments::pay_reward(
 					lane_id,
 					lane_data.relayers,
 					&confirmation_relayer,
 					&received_range,
 				);
-			}
+
+				// update relayers state with actual numbers to compute actual weight below
+				relayers_state.unrewarded_relayer_entries = sp_std::cmp::min(
+					relayers_state.unrewarded_relayer_entries,
+					actually_rewarded_relayers,
+				);
+				relayers_state.total_messages = sp_std::cmp::min(
+					relayers_state.total_messages,
+					received_range
+						.end()
+						.checked_sub(*received_range.start())
+						.and_then(|x| x.checked_add(1))
+						.unwrap_or(MessageNonce::MAX),
```

Comment:
1. If something is wrong with numbers, then we shall use the original `total_messages` instead of some wrong number. Hence the `sp_std::cmp::min(relayers_state.total_messages, MessageNonce::MAX)`, which would return original `relayers_state.total_messages`
2. That's true - we may use something like that. The caveat here is the `unwrap_or(...)`, which may be different in different cases. Sometimes it needs to be zero, sometimes - `MAX`. Let's file an issue for that

Reply:
#2032

Reply:
Oh ok, makes sense. Thanks !